src/gc.rs

```diff
@@ -0,0 +1,229 @@
+    pub(crate) state: Mutex<CollectorState>,
+}
+
+impl Collector {
+    pub(crate) fn new(debug_flags: DebugFlags) -> Self {
+        Self {
+            worklist: Vec::new(),
+            black: true,
+            debug_flags,
+            state: Mutex::new(CollectorState::Ready),
+        }
+    }
+
+    pub(crate) fn current_black(&self) -> bool {
+        self.black
+    }
+
+    pub(crate) fn collect(&mut self) {
+        // First check that no call to collect is active
+        {
+            let mut cstate = self.state.lock().unwrap();
+            match *cstate {
+                CollectorState::Ready => *cstate = CollectorState::Marking,
+                _ => {
+                    // Bail out if we reach this point as the collector has already
+                    // been invoked -- and is running -- on another thread.
+                    return;
+                }
+            }
+        }
+
+        // Register spilling is platform specific. This is implemented in
+        // an assembly stub. The fn to scan the stack is passed as a callback
+        unsafe { spill_registers(self, Collector::scan_stack) }
+
+        self.enter_mark_phase();
+
+        *self.state.lock().unwrap() = CollectorState::FinishedMarking;
+
+        if self.debug_flags.mark_only {
+            return;
+        }
+
+        // Flip the meaning of the mark bit, i.e. if false == Black, then it
+        // becomes false == white. This is a simplification which allows us to
+        // avoid resetting the mark bit for every survived object after
+        // collection. Since we do not implement a marking bitmap and instead
+        // store this mark bit in each object header, this would be a very
+        // expensive operation.
+        self.black = !self.black;
+    }
+
+    /// The worklist is populated with potential GC roots during the stack
+    /// scanning phase. The mark phase then traces through this root-set until
+    /// it finds GC objects. Once found, a GC object is coloured black to
+    /// indicate that it is reachable by the mutator, and is therefore *not* a
+    /// candidate for reclaimation.
+    fn enter_mark_phase(&mut self) {
+        while !self.worklist.is_empty() {
+            let PtrInfo { ptr, size, gc } = self.worklist.pop().unwrap();
+
+            if gc {
+                // For GC objects, the pointer recorded in the alloc metadata
+                // list points to the beginning of the object -- *not* the
+                // object's header. This means that unlike regular allocations,
+                // `ptr` will never point to the beginning of the allocation
+                // block.
+                let obj = unsafe { Gc::from_raw(ptr as *const i8) };
+                if self.colour(obj) == Colour::Black {
+                    break;
+                }
+
+                self.mark(obj, Colour::Black);
+            }
+        }
+    }
+
+    #[no_mangle]
+    extern "sysv64" fn scan_stack(&mut self, rsp: StackAddress) {
```

Comment:
Why are we placing ABI constraints here? Does C code call this?

Reply:
Because this function is called from the `SpillRegisters_x86_64` stub which sets up its arguments according to SYS V.